Add `ECDSAKey` support to `JWKS`

JaapWijnen opened the pull request:
- Allows to create `JWK`'s using `ECDSAKey`s 
- Allows use of `ECDSAKey`s with `JWTSigner`.
- Prevents a crash when using the `.es384` curve for `ECDSAKey`s

```swift
let key = ECDSAKey.generate(curve: .p384)
let jwks = JWKS(keys: [
   JWK.ecdsa(.es384, identifier: JWKIdentifier(string: "token-kid"), x: key.parameters!.x, y: key.parameters!.y, curve: .p384)
])
```

```swift 
let key = ECDSAKey.generate(curve: .p384)
let signer = JWTSigner.es384(key: key)
```

tanner0101 commented on Sources/JWTKit/ECDSA/ECDSAKey.swift:
```diff
@@ -59,8 +59,45 @@ public final class ECDSAKey: OpenSSLKey {
         CJWTKitBoringSSL_EC_KEY_free(self.c)
     }
+
+    public func getParameters() throws -> Parameters {
```
Since this only throws one error instance anyway, I think a computed property and optional is fine here. 

```swift
var parameters: Parameters? { get }
```

tanner0101 commented on Sources/JWTKit/ECDSA/ECDSAKey.swift:
```diff
@@ -59,8 +59,45 @@ public final class ECDSAKey: OpenSSLKey {
         self.c = c
     }
+    
+    convenience init(x: String, y: String, curve: Curve = .p521) throws {
```
What about `public init(parameters:curve:)` using the `Parameters` type below?

tanner0101 commented on Sources/JWTKit/JWK/JWK.swift:
```diff
@@ -4,67 +4,95 @@ import class Foundation.JSONDecoder
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        try container.encode(self.rawValue)
+    }
```
Looks like an indentation error here.

tanner0101 commented on Sources/JWTKit/JWK/JWK.swift:
```diff
@@ -4,67 +4,95 @@ import class Foundation.JSONDecoder
-    /// The `kty` (key type) parameter identifies the cryptographic algorithm
+     
+    ///  The `kty` (key type) parameter identifies the cryptographic algorithm
     ///  family used with the key, such as `RSA` or `EC`. The `kty` value
```
This should only have one space after `///`

tanner0101 commented on Sources/JWTKit/JWK/JWK.swift:
```diff
@@ -79,26 +107,63 @@ public struct JWK: Decodable {
-    
+
+    // ECDSA keys
+    public var x: String?
```
Are there more "technically correct" names than x and y? i.e., `d` in RSA is actually the "private exponent"

JaapWijnen replied:
It doesn't really have a good name as far as I can see. We can call it a `generatorPoint` and rename `Parameters` to `GeneratorPoint` as well as it's a point on the elliptic curve apparently :) How does that sound? But then I'll redo the storing in that way as well instead of x and y separate

JaapWijnen replied:
@tanner0101 I've implemented your other suggestions. Let me know what you think regarding the above!

tanner0101 commented on Sources/JWTKit/JWK/JWK.swift:
```diff
@@ -4,67 +4,95 @@ import class Foundation.JSONDecoder
         case rsa
-        
+        /// EC
+        case ec
```
Is this the technically correct name? I thought it was "ecdsa". We should double check this

tanner0101 commented on Sources/JWTKit/JWK/JWK.swift:
```diff
@@ -79,26 +107,63 @@ public struct JWK: Decodable {
         self = try JSONDecoder().decode(JWK.self, from: Data(json.utf8))
     }
+    
+    public init(rsaKeyUsing algorithm: Algorithm?, keyIdentifier: JWKIdentifier?, n: String?, e: String?, d: String?) {
```
What about static methods here? Let's use the full names of properties here, too.

```swift
static func rsa(_ algorithm:identifier:modulus:...)
```

tanner0101 commented on Sources/JWTKit/Utilities/BN.swift:
```diff
@@ -0,0 +1,45 @@
+//  BN.swift
+//  JWTKit
+//
+//  Created by Alex Tran-Qui on 17/08/2020.
```
rm header

tanner0101 commented on Sources/JWTKit/Utilities/BN.swift:
```diff
@@ -0,0 +1,45 @@
+import Foundation
+import CJWTKitBoringSSL
+
+class BN {
```
We should avoid abbreviations: `BigNumber`.

JaapWijnen commented:
@tanner0101 While testing some more I ran into another bug, and fixed a bug I introduced myself. There is some error I can't seem to fix. I added a test to share the problem. When trying to sign a payload with an ECDSA key I keep running into a `signing algorithm error`. I can't really find what's causing this yet as I'm not super familiar with the SSL libraries. Any ideas?

JaapWijnen commented:
@tanner0101 did some digging on the KeyType naming

> The "kty" (key type) parameter identifies the cryptographic algorithm
   family used with the key, such as "RSA" or "EC".  "kty" values should
   either be registered in the IANA "JSON Web Key Types" registry
   established by [JWA] or be a value that contains a Collision-
   Resistant Name.  The "kty" value is a case-sensitive string.  This
   member MUST be present in a JWK.

https://tools.ietf.org/html/rfc7517#ref-JWA
section 4.1

JaapWijnen commented:
So I renamed both of them, let me know what you think as this could potentially break for users using lowercase (which is against the spec)

tanner0101 commented:
These changes are now available in [4.2.0](https://github.com/vapor/jwt-kit/releases/tag/4.2.0)
